src/hasSubObject.ts

```diff
@@ -9,6 +9,56 @@ type HasSubObjectGuard<T, S> = Simplify<
   Branded<S & T, typeof HAS_SUB_OBJECT_BRAND>
 >;
 
+type HasSubObjectObjectValue<A, B> = Partial<{
+  [Key in keyof A & keyof B]: A[Key] & B[Key] extends never
+    ? B[Key]
+    : A[Key] | B[Key] extends object
+      ? HasSubObjectObjectValue<A[Key], B[Key]>
+      : A[Key] & B[Key] extends object
+        ? B[Key]
+        : A[Key];
+}> & {
+  [Key in
+    | Exclude<keyof A, keyof B>
+    | Exclude<keyof B, keyof A>]: Key extends keyof B ? B[Key] : never;
+};
+
+type HasSubObjectData<
```

Comment:
**nit:** i don't really like this name, but i can't think of a better one right now. same with `HasSubObjectSubObject`

Reply:
@cjquines Alternative is to remove `HasSubObject` prefix and have:

- `HasSubObjectData` -> `Data`
- `HasSubObjectSubObject` -> `SubObject` or `DataSubObject`
- `HasSubObjectObjectValue` -> `DataObjectValue` or `ObjectValue`

Reply:
i'm fine going ahead with this name for now, since we can always change it later